File: python_app/billing_engine.py
import math
from datetime import date

def calculate_bill(
    consumption_kl: float,
    consumer: dict,
    rates: dict,
    previous_outstanding: float = 0.0,
    credit_balance: float = 0.0,
    last_payment_date: date = None,
    payment_date: date = None,
    avg_6month_kl: float = None
) -> dict:
    """Calculates water bill breakdown dynamically using charges config rates."""
    category = consumer.get("category", "Domestic")
    meter_size = consumer.get("meter_size", "15mm")
    status = consumer.get("consumer_status", "Active")
    flat_rate_attr = consumer.get("custom_attributes", {}).get("is_flat_rate_rural", False)
    is_flat_rate_rural = str(flat_rate_attr).lower() == "true" or bool(consumer.get("is_flat_rate_rural", False))
    
    # 1. Determine size type (15-25mm standard vs bulk >25mm)
    is_bulk = meter_size not in ["15mm", "20mm", "25mm"]
    
    # 2. Water charge & slabs calculation
    water_charge = 0.0
    slab_details = []
    min_charge_applied = False
    min_charge_amount = 0.0
    
    # Check special domestic waiver: 15mm, active/functional meter, and consumption <= 15 KL
    is_domestic_15mm_waiver = (
        category == "Domestic" and 
        meter_size == "15mm" and 
        status == "Active" and 
        consumption_kl <= 15.0 and
        not is_flat_rate_rural
    )
    
    if is_flat_rate_rural and category == "Domestic" and meter_size == "15mm":
        # Flat Rate Rural Domestic
        water_charge = float(rates.get("domestic_flat_rural", 110.0))
        slab_details.append({
            "slab": "Flat Rate Rural",
            "kl": consumption_kl,
            "rate": water_charge,
            "amount": water_charge
        })
    elif is_domestic_15mm_waiver:
        # Domestic waiver applies
        water_charge = 0.0
        slab_details.append({
            "slab": "Domestic Waiver (<=15 KL)",
            "kl": consumption_kl,
            "rate": 0.0,
            "amount": 0.0
        })
    elif is_bulk:
        # Bulk connections (>25mm)
        rate_key = f"bulk_{category.lower().replace('-', '')}_rate"
        bulk_rate = float(rates.get(rate_key, 25.0 if category == "Domestic" else (97.0 if category == "Non-Domestic" else 220.0)))
        water_charge = consumption_kl * bulk_rate
        
        slab_details.append({
            "slab": f"Bulk {category}",
            "kl": consumption_kl,
            "rate": bulk_rate,
            "amount": water_charge
        })
        
        # Apply bulk minimum monthly charges
        cat_key = "dom" if category == "Domestic" else ("nondom" if category == "Non-Domestic" else "ind")
        min_key = f"bulk_min_{cat_key}_{meter_size}"
        min_charge_amount = float(rates.get(min_key, 0.0))
        if water_charge < min_charge_amount:
            water_charge = min_charge_amount
            min_charge_applied = True
            
    else:
        # Standard connections (15mm - 25mm) slab calculation
        slab_details = get_slab_breakdown(consumption_kl, category, meter_size, rates)
        water_charge = sum(s["amount"] for s in slab_details)
        
        # Apply standard minimum monthly charges
        # Minimum charges apply only if consumption > 15 KL for Domestic
        if category == "Domestic":
            if consumption_kl > 15.0:
                if meter_size == "15mm":
                    # Check avg_6month_kl or current consumption
                    avg_check = avg_6month_kl if avg_6month_kl is not None else consumption_kl
                    if avg_check <= 8.0:
                        min_charge_amount = float(rates.get("domestic_min_15mm_avg_low", 88.0))
                    else:
                        min_charge_amount = float(rates.get("domestic_min_15mm_avg_high", 220.0))
                elif meter_size == "20mm":
                    min_charge_amount = float(rates.get("domestic_min_20mm", 880.0))
                elif meter_size == "25mm":
                    min_charge_amount = float(rates.get("domestic_min_25mm", 2200.0))
                
                if water_charge < min_charge_amount:
                    water_charge = min_charge_amount
                    min_charge_applied = True
        else:
            # Non-Domestic / Industrial always has minimum monthly charges
            cat_pref = "nondomestic" if category == "Non-Domestic" else "industrial"
            min_key = f"{cat_pref}_min_{meter_size}"
            min_charge_amount = float(rates.get(min_key, 0.0))
            if water_charge < min_charge_amount:
                water_charge = min_charge_amount
                min_charge_applied = True

    # 3. Fixed Charge (Capital Renovation)
    fixed_charge = 0.0
    if is_flat_rate_rural and category == "Domestic" and meter_size == "15mm":
        # Fixed charge not applicable on flat rate rural connections
        fixed_charge = 0.0
    elif is_bulk:
        cat_key = "dom" if category == "Domestic" else ("nondom" if category == "Non-Domestic" else "ind")
        fixed_key = f"bulk_fixed_{cat_key}_{meter_size}"
        fixed_charge = float(rates.get(fixed_key, 0.0))
    else:
        fixed_key = f"fixed_charge_{category.lower().replace('-', '')}"
        fixed_charge = float(rates.get(fixed_key, 0.0))
        
    # 4. Meter Service Charge
    meter_service_charge = 0.0
    if is_bulk:
        svc_key = f"bulk_svc_{meter_size}"
        meter_service_charge = float(rates.get(svc_key, 0.0))
    else:
        svc_key = f"meter_svc_{meter_size}"
        meter_service_charge = float(rates.get(svc_key, 0.0))

    # 5. Infrastructure Development Surcharge (IDS)
    # Up to 15 KL: No IDS
    # 15.001 - 40 KL: 25% of total monthly charges
    # Above 40 KL: 35% of total monthly charges
    # Total monthly charges = water_charge + fixed_charge + meter_service_charge
    ids_rate_pct = 0.0
    if consumption_kl > 15.0 and consumption_kl <= 40.0:
        ids_rate_pct = 25.0
    elif consumption_kl > 40.0:
        ids_rate_pct = 35.0
        
    monthly_subtotal = water_charge + fixed_charge + meter_service_charge
    ids_charge = round((ids_rate_pct / 100.0) * monthly_subtotal, 2)
    
    # 5.1 Sewerage Tax (Clause 14)
    # Sewerage Tax only applies if water charges are chargeable (water_charge > 0)
    # OR if it's Own/Mixed supply (as per Clause 14 Mixed is treated as Own)
    sewerage_tax = 0.0
    has_sewerage = bool(consumer.get("has_sewerage", False))
    supply_type = consumer.get("supply_type", "PHED")
    sub_category = consumer.get("sewerage_sub_category")

    if has_sewerage:
        # Clause 14 Note: If water charges are not chargeable, Sewerage Tax will also not be chargeable.
        # This applies specifically to the Domestic 15mm <= 15 KL waiver.
        if is_domestic_15mm_waiver:
            sewerage_tax = 0.0
        elif supply_type == "PHED":
            # 20% of water charges
            if water_charge > 0:
                sewerage_tax = round(water_charge * (float(rates.get("sewerage_phed_supply_rate_pct", 20.0)) / 100.0), 2)
        else:
            # Own or Mixed Supply (Treated as Own)
            if sub_category == "Hotel":
                rooms = int(consumer.get("rooms_count", 0))
                sewerage_tax = rooms * float(rates.get("sewerage_own_hotel_per_room", 31.25))
            elif sub_category == "Restaurant":
                sewerage_tax = float(rates.get("sewerage_own_restaurant", 200.00))
            elif sub_category == "Cinema":
                sewerage_tax = float(rates.get("sewerage_own_cinema", 400.00))
            elif sub_category == "Car/Truck Service Station":
                sewerage_tax = float(rates.get("sewerage_own_car_service", 200.00))
            elif sub_category == "Scooter Service Station":
                sewerage_tax = float(rates.get("sewerage_own_scooter_service", 62.50))
            elif sub_category == "Other Industrial/Commercial":
                rooms = int(consumer.get("rooms_count", 0))
                sewerage_tax = rooms * float(rates.get("sewerage_own_other_ind_comm_per_room", 12.50))
            elif sub_category == "Domestic":
                sewerage_tax = float(rates.get("sewerage_own_domestic", 12.50))
            elif sub_category == "House > 200sqm":
                area = float(consumer.get("plot_area_sqm", 0.0))
                sewerage_tax = (area / 100.0) * float(rates.get("sewerage_own_house_large_per_100sqm", 6.25))

    # 5.2 Sewerage Treatment Plant (STP) Charges (Clause 15)
    # 13% of water charges
    stp_charge = 0.0
    if bool(consumer.get("has_stp", False)):
        if water_charge > 0:
            stp_charge = round(water_charge * (float(rates.get("stp_charge_rate_pct", 13.0)) / 100.0), 2)

    monthly_subtotal = water_charge + fixed_charge + meter_service_charge + sewerage_tax + stp_charge
    ids_charge = round((ids_rate_pct / 100.0) * monthly_subtotal, 2)

    subtotal_before_lps = monthly_subtotal + ids_charge
    
    # 6. Apply credit balance to new monthly charges
    credit_applied = 0.0
    if credit_balance > 0:
        if credit_balance >= subtotal_before_lps:
            credit_applied = subtotal_before_lps
            remaining_credit = credit_balance - subtotal_before_lps
        else:
            credit_applied = credit_balance
            remaining_credit = 0.0
    else:
        remaining_credit = 0.0
        
    subtotal_after_credit = subtotal_before_lps - credit_applied
    
    # 7. Late Payment Surcharge (LPS)
    lps_amount = 0.0
    lps_type = "none"
    lps_applicable = False
    
    # LPS applies if payment date is set and is after last payment date, 
    # AND there is remaining unpaid subtotal (i.e. credit did not cover it).
    if last_payment_date and payment_date and subtotal_after_credit > 0:
        lps_res = apply_lps(subtotal_before_lps, last_payment_date, payment_date, credit_balance, previous_outstanding)
        lps_amount = lps_res["lps_amount"]
        lps_type = lps_res["lps_type"]
        lps_applicable = lps_res["lps_applicable"]
        
    # 8. Accumulate totals
    total_before_rounding = subtotal_after_credit + previous_outstanding + lps_amount
    total_amount = math.ceil(total_before_rounding)
    
    # 9. Anomaly detection
    is_anomaly = False
    if avg_6month_kl is not None and consumption_kl > (3 * avg_6month_kl):
        is_anomaly = True

    return {
        "water_charge": water_charge,
        "slab_details": slab_details,
        "minimum_charge_applied": min_charge_applied,
        "minimum_charge_amount": min_charge_amount,
        "fixed_charge": fixed_charge,
        "meter_service_charge": meter_service_charge,
        "sewerage_tax": sewerage_tax,
        "stp_charge": stp_charge,
        "ids_charge": ids_charge,
        "ids_rate_pct": ids_rate_pct,
        "previous_outstanding": previous_outstanding,
        "credit_applied": credit_applied,
        "remaining_credit": remaining_credit,
        "lps_amount": lps_amount,
        "lps_type": lps_type,
        "lps_applicable": lps_applicable,
        "subtotal_before_lps": subtotal_before_lps,
        "total_before_rounding": total_before_rounding,
        "total_amount": total_amount,
        "is_anomaly": is_anomaly,
        "is_flat_rate_rural": is_flat_rate_rural
    }

def get_slab_breakdown(consumption_kl: float, category: str, meter_size: str, rates: dict) -> list:
    """Calculates standard slab-by-slab water charges."""
    slab_details = []
    remaining = consumption_kl
    
    if category == "Domestic":
        # Slabs: 0-8 KL (rate 7), 8-15 KL (rate 9), 15-40 KL (rate 18), >40 KL (rate 22)
        slabs = [
            ("Slab 0-8 KL", 8.0, float(rates.get("domestic_slab_a_rate", 7.00))),
            ("Slab 8-15 KL", 7.0, float(rates.get("domestic_slab_b_rate", 9.00))),
            ("Slab 15-40 KL", 25.0, float(rates.get("domestic_slab_c_rate", 18.00))),
            ("Slab Above 40 KL", float("inf"), float(rates.get("domestic_slab_d_rate", 22.00)))
        ]
    elif category == "Non-Domestic":
        # Slabs: 0-15 KL (rate 40), 15-40 KL (rate 73), >40 KL (rate 97)
        slabs = [
            ("Slab 0-15 KL", 15.0, float(rates.get("nondomestic_slab_a_rate", 40.00))),
            ("Slab 15-40 KL", 25.0, float(rates.get("nondomestic_slab_b_rate", 73.00))),
            ("Slab Above 40 KL", float("inf"), float(rates.get("nondomestic_slab_c_rate", 97.00)))
        ]
    elif category == "Industrial":
        # Slabs: 0-15 (rate 154), 15-40 (rate 198), >40 (rate 220)
        slabs = [
            ("Slab 0-15 KL", 15.0, float(rates.get("industrial_slab_a_rate", 154.00))),
            ("Slab 15-40 KL", 25.0, float(rates.get("industrial_slab_b_rate", 198.00))),
            ("Slab Above 40 KL", float("inf"), float(rates.get("industrial_slab_c_rate", 220.00)))
        ]
    else:
        # Fallback to simple domestic
        slabs = [("Slab All", float("inf"), 10.00)]
        
    for name, limit, rate in slabs:
        if remaining <= 0:
            break
        chunk = min(remaining, limit)
        amount = chunk * rate
        slab_details.append({
            "slab": name,
            "kl": chunk,
            "rate": rate,
            "amount": round(amount, 2)
        })
        remaining -= chunk
        
    return slab_details

def apply_lps(bill_total: float, last_payment_date: date, payment_date: date, credit_balance: float, outstanding: float) -> dict:
    """Calculates Late Payment Surcharge (LPS) amount and type based on payment delay."""
    # Check if credit balance covers the full bill. If so, no LPS.
    if credit_balance >= bill_total:
        return {"lps_amount": 0.0, "lps_type": "none", "lps_applicable": False}
        
    if payment_date <= last_payment_date:
        return {"lps_amount": 0.0, "lps_type": "none", "lps_applicable": False}
        
    # Calculate difference in months / days
    from dateutil.relativedelta import relativedelta
    diff = relativedelta(payment_date, last_payment_date)
    months_diff = diff.years * 12 + diff.months
    
    # Subtotal to apply LPS rate to is the monthly bill total
    lps_base = bill_total
    
    # Check if delay is within 2 calendar months after the last_payment_date
    limit_2_months = last_payment_date + relativedelta(months=2)
    
    if payment_date <= limit_2_months:
        lps_amount = round(lps_base * 0.10, 2)
        return {"lps_amount": lps_amount, "lps_type": "10pct", "lps_applicable": True}
    else:
        # Beyond 2 months: 10% LPS + 18% annual interest on outstanding (meaning the past outstanding balance being paid)
        lps_amount_10 = lps_base * 0.10
        
        # Interest is 18% annual on outstanding balance for the period of delay
        # Calculate exact days past due
        days_past_due = (payment_date - last_payment_date).days
        interest = outstanding * 0.18 * (days_past_due / 365.0)
        
        total_lps = round(lps_amount_10 + interest, 2)
        return {"lps_amount": total_lps, "lps_type": "10pct_plus_interest", "lps_applicable": True}

File: public/billing-engine.js
// Web Billing Engine (Rajasthan Water Tariff 2025 parity)

export function parseDate(str) {
  if (!str) return new Date();
  const parts = str.split("-");
  return new Date(parseInt(parts[2]), parseInt(parts[1]) - 1, parseInt(parts[0]));
}

export function formatDate(dateObj) {
  if (!dateObj) return "";
  const d = String(dateObj.getDate()).padStart(2, "0");
  const m = String(dateObj.getMonth() + 1).padStart(2, "0");
  const y = dateObj.getFullYear();
  return `${d}-${m}-${y}`;
}

export function addMonths(date, months) {
  const d = new Date(date.getTime());
  const currentDay = d.getDate();
  d.setDate(1); // Set to 1st to prevent overflow
  d.setMonth(d.getMonth() + months);
  
  // Get last day of the target month
  const lastDay = new Date(d.getFullYear(), d.getMonth() + 1, 0).getDate();
  d.setDate(Math.min(currentDay, lastDay));
  return d;
}

export function formatCurrency(amount) {
  const amt = amount || 0;
  return "₹" + amt.toLocaleString("en-IN", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

export function formatKL(value) {
  const val = value || 0;
  return val.toFixed(2) + " KL";
}

export function getSlabBreakdown(consumptionKL, category, meterSize, rates) {
  const slabDetails = [];
  let remaining = consumptionKL;
  let slabs = [];
  
  if (category === "Domestic") {
    slabs = [
      { name: "Slab 0-8 KL", limit: 8.0, rate: parseFloat(rates.domestic_slab_a_rate || 7.00) },
      { name: "Slab 8-15 KL", limit: 7.0, rate: parseFloat(rates.domestic_slab_b_rate || 9.00) },
      { name: "Slab 15-40 KL", limit: 25.0, rate: parseFloat(rates.domestic_slab_c_rate || 18.00) },
      { name: "Slab Above 40 KL", limit: Infinity, rate: parseFloat(rates.domestic_slab_d_rate || 22.00) }
    ];
  } else if (category === "Non-Domestic") {
    slabs = [
      { name: "Slab 0-15 KL", limit: 15.0, rate: parseFloat(rates.nondomestic_slab_a_rate || 40.00) },
      { name: "Slab 15-40 KL", limit: 25.0, rate: parseFloat(rates.nondomestic_slab_b_rate || 73.00) },
      { name: "Slab Above 40 KL", limit: Infinity, rate: parseFloat(rates.nondomestic_slab_c_rate || 97.00) }
    ];
  } else if (category === "Industrial") {
    slabs = [
      { name: "Slab 0-15 KL", limit: 15.0, rate: parseFloat(rates.industrial_slab_a_rate || 154.00) },
      { name: "Slab 15-40 KL", limit: 25.0, rate: parseFloat(rates.industrial_slab_b_rate || 198.00) },
      { name: "Slab Above 40 KL", limit: Infinity, rate: parseFloat(rates.industrial_slab_c_rate || 220.00) }
    ];
  } else {
    slabs = [{ name: "Slab All", limit: Infinity, rate: 10.00 }];
  }
  
  for (const s of slabs) {
    if (remaining <= 0) break;
    const chunk = Math.min(remaining, s.limit);
    const amount = chunk * s.rate;
    slabDetails.push({
      slab: s.name,
      kl: chunk,
      rate: s.rate,
      amount: Math.round(amount * 100) / 100
    });
    remaining -= chunk;
  }
  return slabDetails;
}

export function applyLPS(billTotal, lastPaymentDate, paymentDate, creditBalance, outstanding) {
  if (creditBalance >= billTotal) {
    return { lps_amount: 0.0, lps_type: "none", lps_applicable: false };
  }
  
  const lastPay = typeof lastPaymentDate === "string" ? parseDate(lastPaymentDate) : lastPaymentDate;
  const payD = typeof paymentDate === "string" ? parseDate(paymentDate) : paymentDate;
  
  if (payD <= lastPay) {
    return { lps_amount: 0.0, lps_type: "none", lps_applicable: false };
  }
  
  // Calculate limit 2 months after lastPay
  const limit2Months = addMonths(lastPay, 2);
  const lpsBase = billTotal;
  
  if (payD <= limit2Months) {
    const lpsAmount = Math.round(lpsBase * 0.10 * 100) / 100;
    return { lps_amount: lpsAmount, lps_type: "10pct", lps_applicable: true };
  } else {
    const lpsAmount10 = lpsBase * 0.10;
    const diffTime = Math.abs(payD - lastPay);
    const daysPastDue = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    const interest = outstanding * 0.18 * (daysPastDue / 365.0);
    const totalLPS = Math.round((lpsAmount10 + interest) * 100) / 100;
    return { lps_amount: totalLPS, lps_type: "10pct_plus_interest", lps_applicable: true };
  }
}

export default function calculateBill(
  consumptionKL,
  consumer,
  rates,
  previousOutstanding = 0,
  creditBalance = 0,
  lastPaymentDate = null,
  paymentDate = null,
  avg6MonthKL = null
) {
  const category = consumer.category || "Domestic";
  const meterSize = consumer.meter_size || "15mm";
  const status = consumer.consumer_status || "Active";
  const customAttrs = consumer.custom_attributes || {};
  const isFlatRateRural = String(customAttrs.is_flat_rate_rural || false).toLowerCase() === "true" || Boolean(consumer.is_flat_rate_rural);
  
  const isBulk = !["15mm", "20mm", "25mm"].includes(meterSize);
  
  let waterCharge = 0.0;
  let slabDetails = [];
  let minChargeApplied = false;
  let minChargeAmount = 0.0;
  
  const isDomestic15mmWaiver = (
    category === "Domestic" &&
    meterSize === "15mm" &&
    status === "Active" &&
    consumptionKL <= 15.0 &&
    !isFlatRateRural
  );
  
  if (isFlatRateRural && category === "Domestic" && meterSize === "15mm") {
    waterCharge = parseFloat(rates.domestic_flat_rural || 110.00);
    slabDetails.push({
      slab: "Flat Rate Rural",
      kl: consumptionKL,
      rate: waterCharge,
      amount: waterCharge
    });
  } else if (isDomestic15mmWaiver) {
    waterCharge = 0.0;
    slabDetails.push({
      slab: "Domestic Waiver (<=15 KL)",
      kl: consumptionKL,
      rate: 0.0,
      amount: 0.0
    });
  } else if (isBulk) {
    const catKey = `bulk_${category.toLowerCase().replace("-", "")}_rate`;
    let defaultRate = 25.0;
    if (category === "Non-Domestic") defaultRate = 97.0;
    if (category === "Industrial") defaultRate = 220.0;
    
    const bulkRate = parseFloat(rates[catKey] || defaultRate);
    waterCharge = consumptionKL * bulkRate;
    
    slabDetails.push({
      slab: `Bulk ${category}`,
      kl: consumptionKL,
      rate: bulkRate,
      amount: waterCharge
    });
    
    const catPrefix = category === "Domestic" ? "dom" : (category === "Non-Domestic" ? "nondom" : "ind");
    const minKey = `bulk_min_${catPrefix}_${meterSize}`;
    minChargeAmount = parseFloat(rates[minKey] || 0.0);
    if (waterCharge < minChargeAmount) {
      waterCharge = minChargeAmount;
      minChargeApplied = true;
    }
  } else {
    slabDetails = getSlabBreakdown(consumptionKL, category, meterSize, rates);
    waterCharge = slabDetails.reduce((sum, s) => sum + s.amount, 0);
    
    if (category === "Domestic") {
      if (consumptionKL > 15.0) {
        if (meterSize === "15mm") {
          const avgCheck = avg6MonthKL !== null ? avg6MonthKL : consumptionKL;
          minChargeAmount = parseFloat(rates[avgCheck <= 8.0 ? "domestic_min_15mm_avg_low" : "domestic_min_15mm_avg_high"] || (avgCheck <= 8 ? 88.00 : 220.00));
        } else if (meterSize === "20mm") {
          minChargeAmount = parseFloat(rates.domestic_min_20mm || 880.00);
        } else if (meterSize === "25mm") {
          minChargeAmount = parseFloat(rates.domestic_min_25mm || 2200.00);
        }
        
        if (waterCharge < minChargeAmount) {
          waterCharge = minChargeAmount;
          minChargeApplied = true;
        }
      }
    } else {
      const catPrefix = category === "Non-Domestic" ? "nondomestic" : "industrial";
      const minKey = `${catPrefix}_min_${meterSize}`;
      minChargeAmount = parseFloat(rates[minKey] || 0.0);
      if (waterCharge < minChargeAmount) {
        waterCharge = minChargeAmount;
        minChargeApplied = true;
      }
    }
  }
  
  // Fixed charges
  let fixedCharge = 0.0;
  if (isFlatRateRural && category === "Domestic" && meterSize === "15mm") {
    fixedCharge = 0.0;
  } else if (isBulk) {
    const catPrefix = category === "Domestic" ? "dom" : (category === "Non-Domestic" ? "nondom" : "ind");
    const fixedKey = `bulk_fixed_${catPrefix}_${meterSize}`;
    fixedCharge = parseFloat(rates[fixedKey] || 0.0);
  } else {
    const fixedKey = `fixed_charge_${category.toLowerCase().replace("-", "")}`;
    fixedCharge = parseFloat(rates[fixedKey] || 0.0);
  }
  
  // Meter service charges
  let meterServiceCharge = 0.0;
  if (isBulk) {
    const svcKey = `bulk_svc_${meterSize}`;
    meterServiceCharge = parseFloat(rates[svcKey] || 0.0);
  } else {
    const svcKey = `meter_svc_${meterSize}`;
    meterServiceCharge = parseFloat(rates[svcKey] || 0.0);
  }
  
  // IDS Surcharge
  let idsRatePct = 0.0;
  if (consumptionKL > 15.0 && consumptionKL <= 40.0) {
    idsRatePct = 25.0;
  } else if (consumptionKL > 40.0) {
    idsRatePct = 35.0;
  }
  
  // Sewerage & STP
  let sewerageTax = 0.0;
  const hasSewerage = Boolean(consumer.has_sewerage);
  const supplyType = consumer.supply_type || "PHED";
  const subCategory = consumer.sewerage_sub_category;

  if (hasSewerage) {
    if (isDomestic15mmWaiver) {
      sewerageTax = 0.0;
    } else if (supplyType === "PHED") {
      if (waterCharge > 0) {
        sewerageTax = Math.round(waterCharge * (parseFloat(rates.sewerage_phed_supply_rate_pct || 20.0) / 100.0) * 100) / 100;
      }
    } else {
      if (subCategory === "Hotel") {
        const rooms = parseInt(consumer.rooms_count || 0);
        sewerageTax = rooms * parseFloat(rates.sewerage_own_hotel_per_room || 31.25);
      } else if (subCategory === "Restaurant") {
        sewerageTax = parseFloat(rates.sewerage_own_restaurant || 200.00);
      } else if (subCategory === "Cinema") {
        sewerageTax = parseFloat(rates.sewerage_own_cinema || 400.00);
      } else if (subCategory === "Car/Truck Service Station") {
        sewerageTax = parseFloat(rates.sewerage_own_car_service || 200.00);
      } else if (subCategory === "Scooter Service Station") {
        sewerageTax = parseFloat(rates.sewerage_own_scooter_service || 62.50);
      } else if (subCategory === "Other Industrial/Commercial") {
        const rooms = parseInt(consumer.rooms_count || 0);
        sewerageTax = rooms * parseFloat(rates.sewerage_own_other_ind_comm_per_room || 12.50);
      } else if (subCategory === "Domestic") {
        sewerageTax = parseFloat(rates.sewerage_own_domestic || 12.50);
      } else if (subCategory === "House > 200sqm") {
        const area = parseFloat(consumer.plot_area_sqm || 0.0);
        sewerageTax = (area / 100.0) * parseFloat(rates.sewerage_own_house_large_per_100sqm || 6.25);
      }
    }
  }

  let stpCharge = 0.0;
  if (Boolean(consumer.has_stp)) {
    if (waterCharge > 0) {
      stpCharge = Math.round(waterCharge * (parseFloat(rates.stp_charge_rate_pct || 13.0) / 100.0) * 100) / 100;
    }
  }

  const monthlySubtotal = waterCharge + fixedCharge + meterServiceCharge + sewerageTax + stpCharge;
  const idsCharge = Math.round((idsRatePct / 100.0) * monthlySubtotal * 100) / 100;
  const subtotalBeforeLps = monthlySubtotal + idsCharge;
  
  // Credit balance logic
  let creditApplied = 0.0;
  let remainingCredit = 0.0;
  if (creditBalance > 0) {
    if (creditBalance >= subtotalBeforeLps) {
      creditApplied = subtotalBeforeLps;
      remainingCredit = creditBalance - subtotalBeforeLps;
    } else {
      creditApplied = creditBalance;
      remainingCredit = 0.0;
    }
  }
  
  const subtotalAfterCredit = subtotalBeforeLps - creditApplied;
  
  // LPS
  let lpsAmount = 0.0;
  let lpsType = "none";
  let lpsApplicable = false;
  if (lastPaymentDate && paymentDate && subtotalAfterCredit > 0) {
    const lpsRes = applyLPS(subtotalBeforeLps, lastPaymentDate, paymentDate, creditBalance, previousOutstanding);
    lpsAmount = lpsRes.lps_amount;
    lpsType = lpsRes.lps_type;
    lpsApplicable = lpsRes.lps_applicable;
  }
  
  const totalBeforeRounding = subtotalAfterCredit + previousOutstanding + lpsAmount;
  const totalAmount = Math.ceil(totalBeforeRounding);
  
  const isAnomaly = avg6MonthKL !== null && consumptionKL > (3 * avg6MonthKL);
  
  return {
    water_charge: waterCharge,
    slab_details: slabDetails,
    minimum_charge_applied: minChargeApplied,
    minimum_charge_amount: minChargeAmount,
    fixed_charge: fixedCharge,
    meter_service_charge: meterServiceCharge,
    sewerage_tax: sewerageTax,
    stp_charge: stpCharge,
    ids_charge: idsCharge,
    ids_rate_pct: idsRatePct,
    previous_outstanding: previousOutstanding,
    credit_applied: creditApplied,
    remaining_credit: remainingCredit,
    lps_amount: lpsAmount,
    lps_type: lpsType,
    lps_applicable: lpsApplicable,
    subtotal_before_lps: subtotalBeforeLps,
    total_before_rounding: totalBeforeRounding,
    total_amount: totalAmount,
    is_anomaly: isAnomaly,
    is_flat_rate_rural: isFlatRateRural
  };
}
